### ISSUE TITLE
Out of memory error while uploading package

### ISSUE THREAD

**Issue by youcefl** [conan-io/conan#235]
While trying to upload a package I get an out of memory error:
```    
Uploading boost/1.55.0@demo/testing
    Traceback (most recent call last):
      File "<string>", line 10, in <module>
      File "<string>", line 6, in run
      File "C:\Users\lasot\workspace\dev-tools\installer\pyinstaller\conan\build\conan\out00-PYZ.pyz\conans.client.command", line 615, in main
      File "C:\Users\lasot\workspace\dev-tools\installer\pyinstaller\conan\build\conan\out00-PYZ.pyz\conans.client.command", line 529, in run
      File "C:\Users\lasot\workspace\dev-tools\installer\pyinstaller\conan\build\conan\out00-PYZ.pyz\conans.client.command", line 437, in upload
      File "C:\Users\lasot\workspace\dev-tools\installer\pyinstaller\conan\build\conan\out00-PYZ.pyz\conans.client.manager", line 296, in upload
      File "C:\Users\lasot\workspace\dev-tools\installer\pyinstaller\conan\build\conan\out00-PYZ.pyz\conans.client.uploader", line 21, in upload_conan
      File "C:\Users\lasot\workspace\dev-tools\installer\pyinstaller\conan\build\conan\out00-PYZ.pyz\conans.client.proxy", line 182, in upload_conan
      File "C:\Users\lasot\workspace\dev-tools\installer\pyinstaller\conan\build\conan\out00-PYZ.pyz\conans.client.remote_manager", line 28, in upload_conan
      File "C:\Users\lasot\workspace\dev-tools\installer\pyinstaller\conan\build\conan\out00-PYZ.pyz\conans.client.remote_manager", line 126, in compress_export_files
      File "C:\Users\lasot\workspace\dev-tools\installer\pyinstaller\conan\build\conan\out00-PYZ.pyz\conans.client.remote_manager", line 144, in compress_files
      File "C:\Users\lasot\workspace\dev-tools\installer\pyinstaller\conan\build\conan\out00-PYZ.pyz\conans.client.remote_manager", line 140, in addfile
      File "C:\Users\lasot\workspace\dev-tools\installer\pyinstaller\conan\build\conan\out00-PYZ.pyz\tarfile", line 2045, in addfile
      File "C:\Users\lasot\workspace\dev-tools\installer\pyinstaller\conan\build\conan\out00-PYZ.pyz\tarfile", line 275, in copyfileobj
      File "C:\Users\lasot\workspace\dev-tools\installer\pyinstaller\conan\build\conan\out00-PYZ.pyz\gzip", line 241, in write
    MemoryError: out of memory
```
This is using the version 0.8.6 installed with the Windows installer.
The workaround I've found is to install Python 2.7 (64-bit) and then use the version of conan installed using pip.

**Comment by memsharded** [comment]
Ups, seems some problem with the binary created by pyinstaller (we are using pyinstaller to create an exe that embeds the python interpreter, so no python is needed). I have checked pyinstaller, to see if there is a problem with it, or the memory limits could be increased, but doesn't seem possible.

Did it always failed in the upload, even for smaller packages? The resulting boost package is large. What are the characteristics of your system? Which windows version? I guess you have enough RAM, but just in case, it could be that the generated executable uses more memory than the python version. 

In any case, it is probably a good idea to keep using the python/pip version. Releases are much smoother, we might be able to do faster releases, etc. Also having python installed in your system is quite convenient, for example conan-package-tools, which are quite useful for package creators and maintainers, require python installed.

If the windows installer is important for you anyway, please tell me so I will further investigate. Thanks for reporting this issue!

**Comment by memsharded** [comment]
Also, we were using an old version (2.1) of pyinstaller, we have just upgraded it to 3.1.1, and next conan 0.9 release will be done with it. Maybe the upgrade fixes this pyinstaller issue.

**Comment by youcefl** [comment]
Thanks for your quick answer and thanks for this fantastic tool (I'm working on making my company use conan as a C++ packages manager).
To answer your questions:
It didn't fail for smaller packages, it worked without problem. It failed with this out of memory error only when I tried to upload Boost. My system is a Windows 8.1 Pro 64-bit with 16GB of RAM.

**Comment by memsharded** [comment]
You are welcome, thank you!!
Yes, then it is clear it is not a machine (lack of RAM) problem. Haven't tested on Win 8, but on Win 7 and 10, but I guess this is not the problem. Lets wait then to see if next release with pyinstaller 3.1 improves this. Cheers!

**Comment by memsharded** [comment]
Hi! Conan 0.9.2 was generated with new pyinstaller 3.1, probably it could be fixed now, if you are able to check if this problem still persists for you, please tell us, thank you!

**Comment by youcefl** [comment]
Ok, now with the 0.9.4 it works for boost but for a bigger package it fails:
```
c:\tmp>conan upload -r conlan2 --all --force llvm/3.7.1@y****/stable
Uploading llvm/3.7.1@yle/stable
Traceback (most recent call last):
  File "<string>", line 10, in <module>
  File "<string>", line 6, in run
  File "conan\conans\client\command.py", line 680, in main
  File "conan\conans\client\command.py", line 593, in run
  File "conan\conans\client\command.py", line 501, in upload
  File "conan\conans\client\manager.py", line 298, in upload
  File "conan\conans\client\uploader.py", line 21, in upload_conan
  File "conan\conans\client\proxy.py", line 188, in upload_conan
  File "conan\conans\client\remote_manager.py", line 27, in upload_conan
  File "conan\conans\util\files.py", line 108, in build_files_set
  File "conan\conans\util\files.py", line 98, in load
MemoryError
conan returned -1
```
I'm using the same machine as above (upgraded to Windows 10 though).
When using the 64-bit version (installed using pip) it also fails but the behavior is different:
```
C:\Python27>conan upload -r conlan2 --all --force llvm/3.7.1@y******/stable
Uploading llvm/3.7.1@yle/stable
Uploading conan_export.tgz
[==========================                        ]                  ERROR: Unable to connect to conlan2=http://************:9****
```
Notice that uploading other (smaller) packages works so it is not a connection problem contrary to what the error message says.

**Comment by memsharded** [comment]
Ok, great that you found a workaround. Actually I think that we should add it to the conan docs? What did you exactly changed for the worker timeout, and up to what value?

We have recently introduced streaming to disk in file downloads, but it is only used now for tools.download(). We might need to do streaming from/to disk for all transfers to try to alleviate these memory problems.

**Comment by youcefl** [comment]
I just added option -t to the command line I use for running the server,
from gunicorn help:
```
  -t INT, --timeout INT
                        Workers silent for more than this many seconds are
                        killed and restarted. [30]
```
I used a big value of 500s but it may have a worked with a smaller value.

**Comment by lasote** [comment]
It seems that #235, #256 and #349 are closely related.

**Comment by monsdar** [comment]
I'm getting MemoryErrors too when trying to upload large packages. Boost is no problem, but the [Qt package](https://www.conan.io/source/Qt/5.6.1-1/osechet/testing) `Qt/5.6.1-1@osechet/testing` causes the following error:

```
Compressing package...                                                Traceback (most recent call last):
  File "C:\Python27\Scripts\conan-script.py", line 9, in <module>
    load_entry_point('conan==0.11.0', 'console_scripts', 'conan')()
  File "c:\Python27\lib\site-packages\conans\conan.py", line 6, in run
    main(sys.argv[1:])
  File "c:\Python27\lib\site-packages\conans\client\command.py", line 767, in main
    error = command.run(args)
  File "c:\Python27\lib\site-packages\conans\client\command.py", line 679, in run
    method(args[0][1:])
  File "c:\Python27\lib\site-packages\conans\client\command.py", line 583, in upload
    args.remote, all_packages=args.all, force=args.force)
  File "c:\Python27\lib\site-packages\conans\client\manager.py", line 368, in upload
    uploader.upload_conan(conan_reference, all_packages=all_packages, force=force)
  File "c:\Python27\lib\site-packages\conans\client\uploader.py", line 26, in upload_conan
    self.upload_package(PackageReference(conan_ref, package_id), index + 1, total)
  File "c:\Python27\lib\site-packages\conans\client\uploader.py", line 35, in upload_package
    self._remote_proxy.upload_package(package_ref)
  File "c:\Python27\lib\site-packages\conans\client\proxy.py", line 215, in upload_package
    result = self._remote_manager.upload_package(package_reference, remote)
  File "c:\Python27\lib\site-packages\conans\client\remote_manager.py", line 38, in upload_package
    the_files = compress_package_files(the_files)
  File "c:\Python27\lib\site-packages\conans\client\remote_manager.py", line 122, in compress_package_files
    return compress_files(files, PACKAGE_TGZ_NAME, excluded=(CONANINFO, CONAN_MANIFEST))
  File "c:\Python27\lib\site-packages\conans\client\remote_manager.py", line 145, in compress_files
    addfile(the_file, info, tgz)
  File "c:\Python27\lib\site-packages\conans\client\remote_manager.py", line 141, in addfile
    tar.addfile(tarinfo=info, fileobj=the_str)
  File "c:\Python27\lib\tarfile.py", line 2045, in addfile
    copyfileobj(fileobj, self.fileobj, tarinfo.size)
  File "c:\Python27\lib\tarfile.py", line 275, in copyfileobj
    dst.write(buf)
  File "c:\Python27\lib\gzip.py", line 241, in write
    self.fileobj.write(self.compress.compress(data))
MemoryError
```

After installing a 64bit version of Python I do not have problems anymore. The thing is that python.org automatically provides the user with a 32bit setup, so a lot of other people will have that problem too.

**Comment by memsharded** [comment]
This issue should be much improved with the upload with streaming from disk: https://github.com/conan-io/conan/pull/508

and download with streaming to disk: https://github.com/conan-io/conan/pull/497

They are already merged in develop, so will be released in next conan 0.13

**Comment by memsharded** [comment]
Released in conan 0.13.1
